src/main.py

```diff
@@ -0,0 +1,119 @@
+# (c) Open Compute Project
+# (c) Google LLC
+# (c) Igor Semenov (goshik92@gmail.com)
+# 
+# Use of this source code is governed by an MIT-style
+# license that can be found in the LICENSE file or at
+# https://opensource.org/licenses/MIT.
+
+import ocptv.output as tv
+import argparse
+import sh
+import re
+import os
+import contextlib
+from memtester_parsing import MemtesterObserver
+
+def main():
+    parser = argparse.ArgumentParser(description="OCP-compliant memtester wrapper.")
+    parser.add_argument("--mt_path", default="memtester",
+        help="Path to memtester executable")
+    parser.add_argument("--mt_args", default="",
+        help="Memtester arguments packed into a string. See `man memtester` for more detail")
+    parser.add_argument("--mt_log_filename", default=None,
+        help="Path to the file that will contain raw memtester output (both STDOUT and STDERR)")
+    args = parser.parse_args()
+
+    run = tv.TestRun(name="memtester", version="1.0")
+    dut = tv.Dut(id=sh.hostid().strip(), name=sh.hostname().strip())
+    observer = MemtesterObserver()
+    
+    # Use ExitStack() to avoid deep nesting
+    with contextlib.ExitStack() as estack:
+        estack.enter_context(run.scope(dut=dut))
+        estack.enter_context((step := run.add_step("run-memtester")).scope())
+
+        bad_addrs = step.start_measurement_series(name="bad_addrs")
+        
+        # Check if a supported version of memtester is installed in the system
+        def version_callback(version, is_known):
+            step.add_log(tv.LogSeverity.INFO, "Memtester v{} was found".format(version))
+            if (not is_known):
+                m = "This version of memtester was not tested. Expect parsing errors"
+                step.add_log(tv.LogSeverity.WARNING, m)
+        observer.callbacks.version_ready = version_callback
+        
+        # Report loop results
+        def loop_callback(loop):
+            tests = loop.failed_tests()
+            if (len(tests) == 0):
+                m = "Loop #{} finished with success".format(loop.index)
+                step.add_log(tv.LogSeverity.INFO, m)
+            else:
+                # Log failed tests
+                names = ", ".join([t.name for t in tests])
+                m = "Loop #{} failed {} tests: {}".format(loop.index, len(tests), names)
+                step.add_error(symptom="loop-failure", message=m)
+                
+                # Record failed addresses
+                for test in tests:
+                    for r in test.result:
+                        bad_addrs.add_measurement(value=r.addr)
+        observer.callbacks.loop_ready = loop_callback
+        
+        # Report individual tests to make long runs more responsive
+        def test_callback(test):
+            if test.passed():
+                step.add_log(tv.LogSeverity.INFO, "Test '{}' passed".format(test.name))
```

Comment:
nitpick; would say here `Memory test {} passed`, to be more explicit what kind of test

Reply:
yeah, ive thought of that. But consider a user who inspects these results in some test executor UI, only looking at the measurements, logs, etc without the name of the test (scrolling down the page for example). I think it would be useful in that case to see "memory".
I'm not gonna insist on this.

Reply:
Done.